clicommand/pipeline_upload.go

```diff
@@ -315,6 +316,13 @@ var PipelineUploadCommand = cli.Command{
 			if err != nil {
 				l.Warn("%s (%s)", err, r)
 
+				if jsonerr := new(json.MarshalerError); errors.As(err, &jsonerr) {
+					l.Error("Unrecoverable error, skipping retries")
+					r.Break()
+
+					return jsonerr
+				}
+
 				// 422 responses will always fail no need to retry
 				if apierr, ok := err.(*api.ErrorResponse); ok && apierr.Response.StatusCode == 422 {
```

Comment:
While this part should also use the `errors.As` pattern as above, there is another PR out there where I have made this change: https://github.com/buildkite/agent/pull/1927

When this is merged, that should have merge conflicts. But I will take care of it.